### DISCUSSION TITLE
Upgrade from 25 to 26 without losing user sessions?

### DISCUSSION THREAD

**Discussion by erwin1** [keycloak/keycloak#35158] [Q&A] [answered]
The upgrade to 26.0.x presents a significant challenge that is difficult to overcome. 
Until now, we have always upgraded to every major version and we are not using any preview features.

The migration guide of 26 (https://www.keycloak.org/docs/latest/upgrading/) states:

> To prevent losing user sessions upgrade to Keycloak 25 first and enable the persistent sessions feature as outlined in the migration guide for Keycloak 25.

The migration guide for 25 states:

> The feature is a preview feature and disabled by default.

So since we don't use preview features, we run 25.0.6 without persistent user sessions.
The question then is, can we still enable persistent user sessions on a running 25.0.6 installations. Reading the migration guide again:

> If this feature is enabled for an existing deployment that is using only the embedded Infinispan for storing sessions, the existing online user and client sessions will not be migrated to the database. It will only affect newly created online user and online client sessions.

It looks we can, so I was thinking this upgrade procedure should work: 
- enable persistent sessions in 25.0.x
- keep that version running until all sessions are persisted (so at least for "SSO Session Max" time)
- then upgrade to 26.0.x

However, there is also this note in the 25 migration guide:

> Enabling this feature later, by a configuration change, can result in an undefined behavior of Keycloak related to sessions if both persisted and non-persisted sessions co-exist. To prevent that, remove all existing online user and client sessions before the first node is started with the feature enabled. This means all Keycloak nodes need to be stopped and, if used, Infinispan remote cache store and embedded Infinispan JDBC persistence need to be cleared.

So my conclusion is this means one cannot upgrade from 25.0.6 to 26.0.x without losing online user sessions? Or am I missing something?
Ironically, upgrading from 24 to (25 to) 26 would not have this problem because the 24-to-25 upgrade has a session migration feature.

Many thanks in advance for any insights on this matter.

**Comment by mhajas** [comment, marked as answer]
Thank you for reaching out and sorry for the confusion.

In my answer I assume you are using external Infinispan for storing sessions and thanks to that sessions survived Keycloak updatea before. If this is not true let me know.

> The question then is, can we still enable persistent user sessions on a running 25.0.6 installations

You can enable the `persistent-user-sessions` feature on a running 25.0.6 but you need to make sure all sessions are removed before first node with persistent user sessions is started by stoping whole Keycloak cluster and cleaning all external Infinispan caches. This basically means that once you upgraded to 25.0.x without enabling the `persistent-user-sessions` feature it is not possible to officially migrate your sessions to `persistent-user-sessions`.

 > enable persistent sessions in 25.0.x
keep that version running until all sessions are persisted (so at least for "SSO Session Max" time)
then upgrade to 26.0.x

As you correctly pointed out in the comment this can result in unexpected behavior so it is not recommended to mix those two storages. This can result in a situation when some request see a session and some not ([more details here](https://github.com/keycloak/keycloak/issues/29561)). 

> So my conclusion is this means one cannot upgrade from 25.0.6 to 26.0.x without losing online user sessions?

Without enabling the `persistent-user-sessions` feature in 25.0.6 this is not possible. btw. this was never officially possible for prior releases as using remote Infinispan was never supported. We saw many people using it and therefore we wanted to provide them an official way using `persistent-user-sessions` feature.

 > Ironically, upgrading from 24 to (25 to) 26 would not have this problem because the 24-to-25 upgrade has a session migration feature.

That is correct. Unfortunately, we weren't able to provide migration for 26 because 26 changed the way how objects are serialized in embedded/external Infinispan and therefore Keycloak 26 is not capable of reading the objects created by Keycloak 25. The only sessions KC 26 understands are the sessions stored in the database. But this should be the last time the sessions are lost as starting from 26 all sessions are stored in the database and therefore the migration will be done the same way as for other entities.




Also, one unofficial note, technically Keycloak 25 contains the code that migrates sessions from remote Infinispan to persistent user sessions, therefore you can try to provoke the migration path by some custom code addition to Keycloak, however, I would not recommend using this in the production as this is not tested and not documented.

What you would need to do is something around this
1. Stop all Keycloak nodes (Keycloak is down during whole migration to not mix Keycloak nodes with different way of storing sessions)
2. Start something like "migrator node" that has feature `persistent-user-sessions` enabled and on startup executes this code: https://github.com/keycloak/keycloak/blob/aa56637addeff660ef4653f0317f419c212cb84b/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java#L983-L1007
3. After the migration is finished stop the "migrator node" so it does not execute the same code again, for example, on restart
4. Start your Keycloak nodes with `persistent-user-sessions` enabled (you can probably also disable remote Infinispan if you used it only to persist sessions across restart, it is still necessary for multi-site setup though)
5. Now you have your cluster with sessions in the database that would survive upgrading to 26

But as I said this procedure is not tested, not documented and on your own risk.

**Reply by erwin1** [reply]
Hello Michal and thank you for your elaborate response!
I guess It's an unfortunate coincidence of circumstances, but now we know this, we can plan accordingly for this upgrade.

Also it's great that after this upgrade to 26 we won't have to deal with this again 👍
